src/ui.js

```diff
@@ -100,20 +104,29 @@ function handleFileSelect(map, evt) {
         modal.addSuccess();
     };
 
+    const handleTrack = async (file, name) => {
+        for (const track of await parseTrack(file, name)) {
+            track.filename = name;
+            tracks.push(track);
+            map.addTrack(track);
+            modal.addSuccess();
+        }
+    };
+
     const handleFile = async file => {
-        const extension = file.name.split('.').pop().toLowerCase();
+        let name = file.name;
         try {
-            switch (extension) {
-                case 'gpx':
-                    return await handleGPX(file);
-                case 'jpg':
-                case 'jpeg':
-                    return await handleImage(file);
-                default:
-                    throw 'Unsupported file format';
+            if (/\.jpe?g$/i.test(name)) {
+                return await handleImage(file);
+            }
+            const contents = /\.gz$/i.test(name) ? await readGz(file) : await readFile(file);
+            name = name.replace(/\.gz$/i, '');
+            if (/\.(gpx|tcx)$/i.test(name)) {
+                return await handleTrack(contents, file.name);
```

Comment:
This feels rather more convoluted than before, because:

- Reading the contents of an image as text doesn't make sense, so handling images should be done before unzipping.
- Gzipped files have a double extension, so `file.name.split('.').pop().toLowerCase()` would not contain the information on the underlying file type.
- `handleTrack` should not have to deal with the type of the file.
- `handleFile` can't be called recursively, because wrapping the unzipped result back into a file object would be hard and inefficient.

Alternatively, handleFile could be rewritten to take the extension, as before, and either
- call `handleImage`,
- call `handleTextFile` with the contents of the unzipped file, as well as the remaining filename, or
- call `handleTextFile` with the contents of the text file, as well as the remaining filename.